datamodel/src/rendering/Shader.cpp

```diff
@@ -0,0 +1,124 @@
+#include <fstream>
+
+#include <rcbe-engine/datamodel/rendering/Shader.hpp>
+
+#include <rcbe-engine/core/gl_extensions.hpp>
+
+#include <boost/log/trivial.hpp>
+
+namespace rcbe::rendering {
+int Shader::MAXIMUM_VERTEX_ATTRIBUTES = 0;
+
+Shader::~Shader() {
+    glDeleteShader(shader_handle_);
+}
+
+Shader::Shader(const core::EnginePath &path, ShaderType type)
+:
+type_(type)
+, file_path_(path)
+{
+    if (!std::filesystem::exists(file_path_)) {
+        throw std::runtime_error(file_path_.string() + " doesn't exist!");
+    }
+
+    if (type_ == ShaderType::unknown) {
+        throw std::runtime_error("Unknown shader type supplied!");
+    }
+
+    if (!MAXIMUM_VERTEX_ATTRIBUTES)
+        glGetIntegerv(GL_MAX_VERTEX_ATTRIBS, &MAXIMUM_VERTEX_ATTRIBUTES);
+
+
+    shader_handle_ = glCreateShader(static_cast<GLenum>(type_));
+}
+
+Shader::Shader(const Shader &other)
+:
+file_path_ { other.file_path_ }
+, type_ { other.type_ }
+{
+    this->shader_handle_ = glCreateShader(static_cast<GLenum>(this->type_));
+    other.deleteShader();
+}
+
+Shader &Shader::operator=(const Shader &other) {
+    this->file_path_ = other.file_path_;
+    this->type_ = other.type_;
+    this->shader_handle_ = glCreateShader(static_cast<GLenum>(this->type_));
+    other.deleteShader();
+    return *this;
+}
+
+Shader::Shader(Shader &&other)
+:
+file_path_ { std::move(other.file_path_) }
+, type_ { std::move(other.type_) }
+{
+    this->shader_handle_ = glCreateShader(static_cast<GLenum>(this->type_));
+    other.deleteShader();
+}
+
+Shader &Shader::operator=(Shader &&other) {
+    this->file_path_ = std::move(other.file_path_);
+    this->type_ = std::move(other.type_);
+    this->shader_handle_ = glCreateShader(static_cast<GLenum>(this->type_));
+    other.deleteShader();
+    return *this;
+}
+
+const Shader::ShaderHandle &Shader::handle() const {
+    return shader_handle_;
+}
+
+const Shader::ShaderHandle &Shader::compile() {
+    if (!shader_handle_)
+        throw std::runtime_error("Shader was created with error!");
+
+    if (!compiled_) {
+        source_code_.clear();
+
+        std::ifstream ifs(file_path_, std::ios::in);
+
+        source_code_ = std::string(std::istreambuf_iterator<char>(ifs), std::istreambuf_iterator<char>());
+        if (source_code_.empty())
+            throw std::runtime_error("Couldn't read shader source code!");
+
+        const auto * scode = source_code_.c_str();
+
+        glShaderSource(shader_handle_, 1, &scode, nullptr);
```

Comment:
Maybe comment this code, to understand what's going on here.

Reply:
Added a comment